src/main/java/slimeknights/tconstruct/library/tools/ToolCore.java

```diff
@@ -142,290 +199,576 @@ public int getHarvestLevel(ItemStack stack, ToolType toolClass, @Nullable Player
-//        throw new TinkerGuiException();
-//      }
-//      catch(TinkerGuiException e) {
-//        // nothing to do
-//      }
-//    }
-//
-//    // return value shouldn't matter since it's never checked
-//    return true;
-//  }
-//
-//  /* Misc */
-//
-//  // elevate to public, used in AOE preview
-//  public RayTraceResult rayTraceExposed(World worldIn, PlayerEntity player, RayTraceContext.FluidMode fluidMode) {
-//    return Item.rayTrace(worldIn, player, fluidMode);
-//  }
-//
-//
-//  @Override
-//  public boolean shouldCauseBlockBreakReset(ItemStack oldStack, ItemStack newStack) {
-//    return shouldCauseReequipAnimation(oldStack, newStack, false);
-//  }
-//
-//  @Override
-//  public boolean shouldCauseReequipAnimation(ItemStack oldStack, @Nonnull ItemStack newStack, boolean slotChanged) {
-//    if(TagUtil.getResetFlag(newStack)) {
-//      TagUtil.setResetFlag(newStack, false);
-//      return true;
-//    }
-//    if(oldStack == newStack) {
-//      return false;
-//    }
-//    if(slotChanged) {
-//      return true;
-//    }
-//
-//    if(oldStack.hasEffect() != newStack.hasEffect()) {
-//      return true;
-//    }
-//
-//    Multimap<String, AttributeModifier> attributesNew = newStack.getAttributeModifiers(EntityEquipmentSlot.MAINHAND);
-//    Multimap<String, AttributeModifier> attributesOld = oldStack.getAttributeModifiers(EntityEquipmentSlot.MAINHAND);
-//
-//    if(attributesNew.size() != attributesOld.size()) {
-//      return true;
-//    }
-//    for(String key : attributesOld.keySet()) {
-//      if(!attributesNew.containsKey(key)) {
-//        return true;
-//      }
-//      Iterator<AttributeModifier> iter1 = attributesNew.get(key).iterator();
-//      Iterator<AttributeModifier> iter2 = attributesOld.get(key).iterator();
-//      while(iter1.hasNext() && iter2.hasNext()) {
-//        if(!iter1.next().equals(iter2.next())) {
-//          return true;
-//        }
-//      }
-//    }
-//
-//    if(oldStack.getItem() == newStack.getItem() && newStack.getItem() instanceof ToolCore) {
-//      return !isEqualTinkersItem(oldStack, newStack);
-//    }
-//    return !ItemStack.areItemStacksEqual(oldStack, newStack);
-//  }
+    List<IMaterial> materials = toolData.getMaterials();
+    List<PartMaterialRequirement> components = this.getToolDefinition().getRequiredComponents();
+
+    if (materials.size() < components.size()) {
+      return;
+    }
+
+    for (int i = 0; i < components.size(); i++) {
+      PartMaterialRequirement requirement = components.get(i);
+      IMaterial material = materials.get(i);
+
+      Item toolPart = requirement.getPart().get();
+
+      if (toolPart instanceof IMaterialItem) {
+        ItemStack partStack = ((IMaterialItem) toolPart).getItemstackWithMaterial(material);
+
+        tooltips.add((new StringTextComponent("")).append(partStack.getDisplayName()).mergeStyle(TextFormatting.UNDERLINE).modifyStyle(style -> style.setColor(material.getColor())));
```

Comment:
Ditch the `new StringTextComponent` thing here, switch to `copyRaw`. I did not fully understand the formattable stuff at first. Same below